[clang][analyzer] Possible false positive from unix.BlockInCriticalSection

Clang-tidy generates a false positive for the `unix.BlockInCriticalSection` check when a thread that enters a critical section performs a  `recv()` with a non-blocking socket. This is because clang-tidy assumes `fcntl` (system call to set the socket to be non-blocking) to always fail.

Example: https://github.com/dmlc/xgboost/blob/0def8e0bae41e0eeeb4ee078e6e699ad6bef1986/src/collective/loop.cc#L88-L105

<details>
  <summary>Error message</summary>

```
/workspace/include/xgboost/collective/socket.h:696:12: warning: Call to blocking function 'recv' inside of critical section [clang-analyzer-unix.BlockInCriticalSection]
  696 |     return recv(handle_, _buf, len, flags);
      |            ^
/workspace/src/collective/loop.cc:272:5: note: Calling 'Loop::Process'
  272 |     this->Process();
      |     ^~~~~~~~~~~~~~~
/workspace/src/collective/loop.cc:154:3: note: Loop condition is true.  Entering loop body
  154 |   while (true) {
      |   ^
/workspace/src/collective/loop.cc:156:24: note: Calling constructor for 'unique_lock<std::mutex>'
  156 |       std::unique_lock lock{mu_};
      |                        ^~~~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_lock.h:69:2: note: Entering critical section here
   69 |         lock();
      |         ^~~~~~
/workspace/src/collective/loop.cc:156:24: note: Returning from constructor for 'unique_lock<std::mutex>'
  156 |       std::unique_lock lock{mu_};
      |                        ^~~~~~~~~
/workspace/src/collective/loop.cc:167:11: note: Assuming field 'stop_' is false
  167 |       if (stop_) {
      |           ^~~~~
/workspace/src/collective/loop.cc:167:7: note: Taking false branch
  167 |       if (stop_) {
      |       ^
/workspace/src/collective/loop.cc:174:14: note: Assuming the condition is false
  174 |       while (!queue_.empty()) {
      |              ^~~~~~~~~~~~~~~
/workspace/src/collective/loop.cc:174:7: note: Loop condition is false. Execution continues on line 179
  174 |       while (!queue_.empty()) {
      |       ^
/workspace/src/collective/loop.cc:182:17: note: Calling 'Loop::ProcessQueue'
  182 |       auto rc = this->ProcessQueue(&qcopy);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/src/collective/loop.cc:30:7: note: Assuming field 'stop_' is false
   30 |   if (stop_) {
      |       ^~~~~
/workspace/src/collective/loop.cc:30:3: note: Taking false branch
   30 |   if (stop_) {
      |   ^
/workspace/src/collective/loop.cc:38:10: note: Assuming the condition is true
   38 |   while (!qcopy.empty()) {
      |          ^~~~~~~~~~~~~~
/workspace/src/collective/loop.cc:38:3: note: Loop condition is true.  Entering loop body
   38 |   while (!qcopy.empty()) {
      |   ^
/workspace/src/collective/loop.cc:43:29: note: Assuming 'i' is < 'n_ops'
   43 |     for (std::size_t i = 0; i < n_ops; ++i) {
      |                             ^~~~~~~~~
/workspace/src/collective/loop.cc:43:5: note: Loop condition is true.  Entering loop body
   43 |     for (std::size_t i = 0; i < n_ops; ++i) {
      |     ^
/workspace/src/collective/loop.cc:47:7: note: Control jumps to 'case kSleep:'  at line 56
   47 |       switch (op.code) {
      |       ^
/workspace/src/collective/loop.cc:57:11: note:  Execution continues on line 65
   57 |           break;
      |           ^
/workspace/src/collective/loop.cc:43:29: note: Assuming 'i' is >= 'n_ops'
   43 |     for (std::size_t i = 0; i < n_ops; ++i) {
      |                             ^~~~~~~~~
/workspace/src/collective/loop.cc:43:5: note: Loop condition is false. Execution continues on line 69
   43 |     for (std::size_t i = 0; i < n_ops; ++i) {
      |     ^
/workspace/src/collective/loop.cc:70:9: note: Assuming the condition is false
   70 |     if (!poll.fds.empty()) {
      |         ^~~~~~~~~~~~~~~~~
/workspace/src/collective/loop.cc:70:5: note: Taking false branch
   70 |     if (!poll.fds.empty()) {
      |     ^
/workspace/src/collective/loop.cc:80:11: note: Assuming the condition is true
   80 |     CHECK(!qcopy.empty());
      |           ^
/workspace/include/xgboost/logging.h:144:24: note: expanded from macro 'CHECK'
  144 |   if (XGBOOST_EXPECT(!(cond), false))                   \
      |                        ^~~~
/workspace/include/xgboost/base.h:53:54: note: expanded from macro 'XGBOOST_EXPECT'
   53 | #define XGBOOST_EXPECT(cond, ret)  __builtin_expect((cond), (ret))
      |                                                      ^~~~
/workspace/src/collective/loop.cc:80:5: note: Taking false branch
   80 |     CHECK(!qcopy.empty());
      |     ^
/workspace/include/xgboost/logging.h:144:3: note: expanded from macro 'CHECK'
  144 |   if (XGBOOST_EXPECT(!(cond), false))                   \
      |   ^
/workspace/src/collective/loop.cc:83:5: note: Loop condition is true.  Entering loop body
   83 |     for (std::size_t i = 0; i < n_ops; ++i) {
      |     ^
/workspace/src/collective/loop.cc:88:11: note: Assuming field 'sock' is non-null
   88 |       if (!op.sock) {
      |           ^~~~~~~~
/workspace/src/collective/loop.cc:88:7: note: Taking false branch
   88 |       if (!op.sock) {
      |       ^
/workspace/src/collective/loop.cc:91:9: note: Assuming the condition is false
   91 |         CHECK(op.sock->NonBlocking());
      |         ^
/workspace/include/xgboost/logging.h:144:22: note: expanded from macro 'CHECK'
  144 |   if (XGBOOST_EXPECT(!(cond), false))                   \
      |       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
/workspace/include/xgboost/base.h:53:54: note: expanded from macro 'XGBOOST_EXPECT'
   53 | #define XGBOOST_EXPECT(cond, ret)  __builtin_expect((cond), (ret))
      |                                                      ^~~~
/workspace/src/collective/loop.cc:91:9: note: Taking false branch
   91 |         CHECK(op.sock->NonBlocking());
      |         ^
/workspace/include/xgboost/logging.h:144:3: note: expanded from macro 'CHECK'
  144 |   if (XGBOOST_EXPECT(!(cond), false))                   \
      |   ^
/workspace/src/collective/loop.cc:94:7: note: Control jumps to 'case kRead:'  at line 95
   94 |       switch (op.code) {
      |       ^
/workspace/src/collective/loop.cc:96:11: note: Taking true branch
   96 |           if (poll.CheckRead(*op.sock)) {
      |           ^
/workspace/src/collective/loop.cc:97:28: note: Calling 'TCPSocket::Recv'
   97 |             n_bytes_done = op.sock->Recv(op.ptr + op.off, op.n - op.off);
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/include/xgboost/collective/socket.h:696:12: note: Call to blocking function 'recv' inside of critical section
  696 |     return recv(handle_, _buf, len, flags);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 --------------------------------
```

</details>

Suggested fix: Add a note in the documentation about possible false positives with non-blocking sockets.

--- a/clang/include/clang/StaticAnalyzer/Core/PathSensitive/CheckerHelpers.h
+++ b/clang/include/clang/StaticAnalyzer/Core/PathSensitive/CheckerHelpers.h
@@ -6,7 +6,7 @@
 //
 //===----------------------------------------------------------------------===//
 //
-//  This file defines CheckerVisitor.
+//  This file defines various utilities used by checkers.
 //
 //===----------------------------------------------------------------------===//
 
@@ -113,6 +113,10 @@ OperatorKind operationKindFromOverloadedOperator(OverloadedOperatorKind OOK,
                                                  bool IsBinary);
 
 std::optional<SVal> getPointeeVal(SVal PtrSVal, ProgramStateRef State);
+
+/// Returns true if declaration \p D is in std namespace or any nested namespace
+/// or class scope.
+bool isWithinStdNamespace(const Decl *D);
 
 } // namespace ento
 

--- a/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
@@ -245,8 +245,10 @@ static const MemRegion *getRegion(const CallEvent &Call,
                                   const MutexDescriptor &Descriptor,
                                   bool IsLock) {
   return std::visit(
-      [&Call, IsLock](auto &&Descriptor) {
-        return Descriptor.getRegion(Call, IsLock);
+      [&Call, IsLock](auto &Descr) -> const MemRegion * {
+        if (const MemRegion *Reg = Descr.getRegion(Call, IsLock))
+          return Reg->getBaseRegion();
+        return nullptr;
       },
       Descriptor);
 }

--- a/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
@@ -241,13 +241,20 @@ BlockInCriticalSectionChecker::checkDescriptorMatch(const CallEvent &Call,
   return std::nullopt;
 }
 
+static const MemRegion *skipBaseClassRegion(const MemRegion *Reg) {
+  while (const auto *BaseClassRegion = dyn_cast<CXXBaseObjectRegion>(Reg)) {
+    Reg = BaseClassRegion->getSuperRegion();
+  }
+  return Reg;
+}
+
 static const MemRegion *getRegion(const CallEvent &Call,
                                   const MutexDescriptor &Descriptor,
                                   bool IsLock) {
   return std::visit(
       [&Call, IsLock](auto &Descr) -> const MemRegion * {
         if (const MemRegion *Reg = Descr.getRegion(Call, IsLock))
-          return Reg->getBaseRegion();
+          return skipBaseClassRegion(Reg);
         return nullptr;
       },
       Descriptor);

--- a/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
@@ -20,6 +20,7 @@
 #include "clang/StaticAnalyzer/Core/PathSensitive/CallDescription.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/CallEvent.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/CheckerContext.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/CheckerHelpers.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/ProgramStateTrait.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/ProgramState_Fwd.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/SVals.h"
@@ -244,7 +245,7 @@ BlockInCriticalSectionChecker::checkDescriptorMatch(const CallEvent &Call,
 static const MemRegion *skipStdBaseClassRegion(const MemRegion *Reg) {
   while (Reg) {
     const auto *BaseClassRegion = dyn_cast<CXXBaseObjectRegion>(Reg);
-    if (!BaseClassRegion || !BaseClassRegion->getDecl()->isInStdNamespace())
+    if (!BaseClassRegion || !isWithinStdNamespace(BaseClassRegion->getDecl()))
       break;
     Reg = BaseClassRegion->getSuperRegion();
   }
@@ -256,9 +257,7 @@ static const MemRegion *getRegion(const CallEvent &Call,
                                   bool IsLock) {
   return std::visit(
       [&Call, IsLock](auto &Descr) -> const MemRegion * {
-        if (const MemRegion *Reg = Descr.getRegion(Call, IsLock))
-          return skipStdBaseClassRegion(Reg);
-        return nullptr;
+        return skipStdBaseClassRegion(Descr.getRegion(Call, IsLock));
       },
       Descriptor);
 }

--- a/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
@@ -241,10 +241,14 @@ BlockInCriticalSectionChecker::checkDescriptorMatch(const CallEvent &Call,
   return std::nullopt;
 }
 
-static const MemRegion *skipBaseClassRegion(const MemRegion *Reg) {
-  while (const auto *BaseClassRegion = dyn_cast<CXXBaseObjectRegion>(Reg)) {
+static const MemRegion *skipStdBaseClassRegion(const MemRegion *Reg) {
+  do {
+    assert(Reg);
+    const auto *BaseClassRegion = dyn_cast<CXXBaseObjectRegion>(Reg);
+    if (!BaseClassRegion || !BaseClassRegion->getDecl()->isInStdNamespace())
+      break;
     Reg = BaseClassRegion->getSuperRegion();
-  }
+  } while (true);
   return Reg;
 }
 
@@ -254,7 +258,7 @@ static const MemRegion *getRegion(const CallEvent &Call,
   return std::visit(
       [&Call, IsLock](auto &Descr) -> const MemRegion * {
         if (const MemRegion *Reg = Descr.getRegion(Call, IsLock))
-          return skipBaseClassRegion(Reg);
+          return skipStdBaseClassRegion(Reg);
         return nullptr;
       },
       Descriptor);

--- a/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
+++ b/clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp
@@ -242,13 +242,12 @@ BlockInCriticalSectionChecker::checkDescriptorMatch(const CallEvent &Call,
 }
 
 static const MemRegion *skipStdBaseClassRegion(const MemRegion *Reg) {
-  do {
-    assert(Reg);
+  while (Reg) {
     const auto *BaseClassRegion = dyn_cast<CXXBaseObjectRegion>(Reg);
     if (!BaseClassRegion || !BaseClassRegion->getDecl()->isInStdNamespace())
       break;
     Reg = BaseClassRegion->getSuperRegion();
-  } while (true);
+  }
   return Reg;
 }
 

--- a/clang/lib/StaticAnalyzer/Core/CallEvent.cpp
+++ b/clang/lib/StaticAnalyzer/Core/CallEvent.cpp
@@ -38,6 +38,7 @@
 #include "clang/CrossTU/CrossTranslationUnit.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/CallDescription.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/CheckerContext.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/CheckerHelpers.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/DynamicType.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/DynamicTypeInfo.h"
 #include "clang/StaticAnalyzer/Core/PathSensitive/MemRegion.h"
@@ -921,17 +922,6 @@ SVal AnyCXXConstructorCall::getCXXThisVal() const {
   if (Data)
     return loc::MemRegionVal(static_cast<const MemRegion *>(Data));
   return UnknownVal();
-}
-
-static bool isWithinStdNamespace(const Decl *D) {
-  const DeclContext *DC = D->getDeclContext();
-  while (DC) {
-    if (const auto *NS = dyn_cast<NamespaceDecl>(DC);
-        NS && NS->isStdNamespace())
-      return true;
-    DC = DC->getParent();
-  }
-  return false;
 }
 
 void AnyCXXConstructorCall::getExtraInvalidatedValues(ValueList &Values,

--- a/clang/lib/StaticAnalyzer/Core/CheckerHelpers.cpp
+++ b/clang/lib/StaticAnalyzer/Core/CheckerHelpers.cpp
@@ -190,5 +190,16 @@ std::optional<SVal> getPointeeVal(SVal PtrSVal, ProgramStateRef State) {
   return std::nullopt;
 }
 
+bool isWithinStdNamespace(const Decl *D) {
+  const DeclContext *DC = D->getDeclContext();
+  while (DC) {
+    if (const auto *NS = dyn_cast<NamespaceDecl>(DC);
+        NS && NS->isStdNamespace())
+      return true;
+    DC = DC->getParent();
+  }
+  return false;
+}
+
 } // namespace ento
 } // namespace clang

--- a/clang/test/Analysis/block-in-critical-section-inheritance.cpp
+++ b/clang/test/Analysis/block-in-critical-section-inheritance.cpp
@@ -29,3 +29,13 @@ void gh_99628() {
   // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
   m.unlock();
 }
+
+void no_false_positive_gh_104241() {
+  std::mutex m;
+  m.lock();
+  // If inheritance not handled properly, this unlock might not match the lock
+  // above because technically they act on different memory regions:
+  // __mutex_base and mutex.
+  m.unlock();
+  sleep(10); // no-warning
+}

--- a/clang/test/Analysis/block-in-critical-section-inheritance.cpp
+++ b/clang/test/Analysis/block-in-critical-section-inheritance.cpp
@@ -128,3 +128,26 @@ void two_custom_mutex_bases_casts_tn(MyMutex &m) {
   sleep(10);
 }
 
+struct MutexVirtBase1 : virtual std::mutex {
+  void lock1() { lock(); }
+  void unlock1() { unlock(); }
+};
+
+struct MutexVirtBase2 : virtual std::mutex {
+  void lock2() { lock(); }
+  void unlock2() { unlock(); }
+};
+
+struct CombinedVirtMutex : MutexVirtBase1, MutexVirtBase2 {};
+
+void virt_inherited_mutexes_same_base_tn1(CombinedVirtMutex &cvt) {
+  cvt.lock1();
+  cvt.unlock1();
+  sleep(10);
+}
+
+void virt_inherited_mutexes_different_bases_tn(CombinedVirtMutex &cvt) {
+  cvt.lock1();
+  cvt.unlock2(); // Despite a different name, unlock2 acts on the same mutex as lock1
+  sleep(10);
+}

--- a/clang/test/Analysis/block-in-critical-section-inheritance.cpp
+++ b/clang/test/Analysis/block-in-critical-section-inheritance.cpp
@@ -39,3 +39,17 @@ void no_false_positive_gh_104241() {
   m.unlock();
   sleep(10); // no-warning
 }
+
+struct TwoMutexes {
+  std::mutex m1;
+  std::mutex m2;
+};
+
+void two_mutexes_false_negative(TwoMutexes &tm) {
+  tm.m1.lock();
+  tm.m2.unlock();
+  // Critical section is associated with tm now so tm.m1 and tm.m2 are
+  // undistinguishiable
+  sleep(10); // False-negative
+  tm.m1.unlock();
+}

--- a/clang/test/Analysis/block-in-critical-section-inheritance.cpp
+++ b/clang/test/Analysis/block-in-critical-section-inheritance.cpp
@@ -45,11 +45,12 @@ struct TwoMutexes {
   std::mutex m2;
 };
 
-void two_mutexes_false_negative(TwoMutexes &tm) {
+void two_mutexes_no_false_negative(TwoMutexes &tm) {
   tm.m1.lock();
+  // expected-note@-1 {{Entering critical section here}}
   tm.m2.unlock();
-  // Critical section is associated with tm now so tm.m1 and tm.m2 are
-  // undistinguishiable
-  sleep(10); // False-negative
+  sleep(10);
+  // expected-warning@-1 {{Call to blocking function 'sleep' inside of critical section}}
+  // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
   tm.m1.unlock();
 }

--- a/clang/test/Analysis/block-in-critical-section-inheritance.cpp
+++ b/clang/test/Analysis/block-in-critical-section-inheritance.cpp
@@ -132,6 +132,7 @@ void two_custom_mutex_bases_casts_tn(MyMutex &m) {
 
 struct MutexVirtBase1 : virtual std::mutex {
   void lock1() { lock(); }
+  // expected-note@-1 {{Entering critical section here}}
   void unlock1() { unlock(); }
 };
 
@@ -153,6 +154,17 @@ void virt_inherited_mutexes_different_bases_tn(CombinedVirtMutex &cvt) {
   cvt.unlock2(); // Despite a different name, unlock2 acts on the same mutex as lock1
   sleep(10);
 }
+
+void virt_inherited_mutexes_different_bases_tp(CombinedVirtMutex &cvt) {
+  cvt.lock1();
+  // expected-note@-1 {{Calling 'MutexVirtBase1::lock1'}}
+  // expected-note@-2 {{Returning from 'MutexVirtBase1::lock1'}}
+  sleep(10);
+  // expected-warning@-1 {{Call to blocking function 'sleep' inside of critical section}}
+  // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
+  cvt.unlock1();
+}
+
 namespace std {
 template <class... MutexTypes> struct scoped_lock {
   explicit scoped_lock(MutexTypes&... m);

--- a/clang/test/Analysis/block-in-critical-section-inheritance.cpp
+++ b/clang/test/Analysis/block-in-critical-section-inheritance.cpp
@@ -57,6 +57,7 @@ void two_mutexes_no_false_negative(TwoMutexes &tm) {
 
 struct MyMutexBase1 : std::mutex {
   void lock1() { lock(); }
+  // expected-note@-1 {{Entering critical section here}}
   void unlock1() { unlock(); }
 };
 struct MyMutexBase2 : std::mutex {
@@ -96,13 +97,14 @@ void custom_mutex_cast_tn(MyMutexBase1 &mb) {
   sleep(10);
 }
 
-void two_custom_mutex_bases_fn(MyMutex &m) {
+void two_custom_mutex_bases_tp(MyMutex &m) {
   m.lock1();
+  // expected-note@-1 {{Calling 'MyMutexBase1::lock1'}}
+  // expected-note@-2 {{Returning from 'MyMutexBase1::lock1'}}
   m.unlock2();
-  // The critical section is associated with `m` ignoring the fact that m holds
-  // two mutexes. hense unlock2 is considered to unlock the same mutex as lock1
-  // locks
-  sleep(10); // False negative
+  sleep(10);
+  // expected-warning@-1 {{Call to blocking function 'sleep' inside of critical section}}
+  // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
   m.unlock1();
 }
 
@@ -112,13 +114,13 @@ void two_custom_mutex_bases_tn(MyMutex &m) {
   sleep(10);
 }
 
-void two_custom_mutex_bases_casts_fn(MyMutex &m) {
+void two_custom_mutex_bases_casts_tp(MyMutex &m) {
   static_cast<MyMutexBase1&>(m).lock();
+  // expected-note@-1 {{Entering critical section here}}
   static_cast<MyMutexBase2&>(m).unlock();
-  // The critical section is associated with `m` ignoring the fact that m holds
-  // two mutexes. hense MyMutexBase2::unlock is considered to unlock the same
-  // mutex as MyMutexBase1::lock locks
-  sleep(10); // False negative
+  sleep(10);
+  // expected-warning@-1 {{Call to blocking function 'sleep' inside of critical section}}
+  // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
   static_cast<MyMutexBase1&>(m).unlock();
 }
 

--- a/clang/test/Analysis/block-in-critical-section-inheritance.cpp
+++ b/clang/test/Analysis/block-in-critical-section-inheritance.cpp
@@ -151,3 +151,32 @@ void virt_inherited_mutexes_different_bases_tn(CombinedVirtMutex &cvt) {
   cvt.unlock2(); // Despite a different name, unlock2 acts on the same mutex as lock1
   sleep(10);
 }
+namespace std {
+template <class... MutexTypes> struct scoped_lock {
+  explicit scoped_lock(MutexTypes&... m);
+  ~scoped_lock();
+};
+template <class MutexType> class scoped_lock<MutexType> {
+public:
+  explicit scoped_lock(MutexType& m) : m(m) { m.lock(); }
+  ~scoped_lock() { m.unlock(); }
+private:
+  MutexType& m;
+};
+} // namespace std
+
+namespace gh_104241 {
+int magic_number;
+std::mutex m;
+
+void fixed() {
+  int current;
+  for (int items_processed = 0; items_processed < 100; ++items_processed) {
+    {
+      std::scoped_lock<std::mutex> guard(m);
+      current = magic_number;
+    }
+    sleep(current); // expected no warning
+  }
+}
+} // namespace gh_104241

--- a/clang/test/Analysis/block-in-critical-section-inheritance.cpp
+++ b/clang/test/Analysis/block-in-critical-section-inheritance.cpp
@@ -54,3 +54,77 @@ void two_mutexes_no_false_negative(TwoMutexes &tm) {
   // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
   tm.m1.unlock();
 }
+
+struct MyMutexBase1 : std::mutex {
+  void lock1() { lock(); }
+  void unlock1() { unlock(); }
+};
+struct MyMutexBase2 : std::mutex {
+  void lock2() { lock(); }
+  void unlock2() { unlock(); }
+};
+struct MyMutex : MyMutexBase1, MyMutexBase2 {};
+// MyMutex has two distinct std::mutex as base classes
+
+void custom_mutex_tp(MyMutexBase1 &mb) {
+  mb.lock();
+  // expected-note@-1 {{Entering critical section here}}
+  sleep(10);
+  // expected-warning@-1 {{Call to blocking function 'sleep' inside of critical section}}
+  // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
+  mb.unlock();
+}
+
+void custom_mutex_tn(MyMutexBase1 &mb) {
+  mb.lock();
+  mb.unlock();
+  sleep(10);
+}
+
+void custom_mutex_cast_tp(MyMutexBase1 &mb) {
+  static_cast<std::mutex&>(mb).lock();
+  // expected-note@-1 {{Entering critical section here}}
+  sleep(10);
+  // expected-warning@-1 {{Call to blocking function 'sleep' inside of critical section}}
+  // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
+  static_cast<std::mutex&>(mb).unlock();
+}
+
+void custom_mutex_cast_tn(MyMutexBase1 &mb) {
+  static_cast<std::mutex&>(mb).lock();
+  static_cast<std::mutex&>(mb).unlock();
+  sleep(10);
+}
+
+void two_custom_mutex_bases_fn(MyMutex &m) {
+  m.lock1();
+  m.unlock2();
+  // The critical section is associated with `m` ignoring the fact that m holds
+  // two mutexes. hense unlock2 is considered to unlock the same mutex as lock1
+  // locks
+  sleep(10); // False negative
+  m.unlock1();
+}
+
+void two_custom_mutex_bases_tn(MyMutex &m) {
+  m.lock1();
+  m.unlock1();
+  sleep(10);
+}
+
+void two_custom_mutex_bases_casts_fn(MyMutex &m) {
+  static_cast<MyMutexBase1&>(m).lock();
+  static_cast<MyMutexBase2&>(m).unlock();
+  // The critical section is associated with `m` ignoring the fact that m holds
+  // two mutexes. hense MyMutexBase2::unlock is considered to unlock the same
+  // mutex as MyMutexBase1::lock locks
+  sleep(10); // False negative
+  static_cast<MyMutexBase1&>(m).unlock();
+}
+
+void two_custom_mutex_bases_casts_tn(MyMutex &m) {
+  static_cast<MyMutexBase1&>(m).lock();
+  static_cast<MyMutexBase1&>(m).unlock();
+  sleep(10);
+}
+

--- a/clang/test/Analysis/block-in-critical-section-nested-namespace.cpp
+++ b/clang/test/Analysis/block-in-critical-section-nested-namespace.cpp
@@ -1,0 +1,42 @@
+
+// RUN: %clang_analyze_cc1 \
+// RUN:   -analyzer-checker=unix.BlockInCriticalSection \
+// RUN:   -std=c++11 \
+// RUN:   -analyzer-output text \
+// RUN:   -verify %s
+
+unsigned int sleep(unsigned int seconds) {return 0;}
+namespace std {
+namespace __detail {
+class __mutex_base {
+public:
+  void lock();
+};
+} // namespace __detail
+
+class mutex : public __detail::__mutex_base{
+public:
+  void unlock();
+  bool try_lock();
+};
+} // namespace std
+
+void gh_99628() {
+  std::mutex m;
+  m.lock();
+  // expected-note@-1 {{Entering critical section here}}
+  sleep(10);
+  // expected-warning@-1 {{Call to blocking function 'sleep' inside of critical section}}
+  // expected-note@-2 {{Call to blocking function 'sleep' inside of critical section}}
+  m.unlock();
+}
+
+void no_false_positive_gh_104241() {
+  std::mutex m;
+  m.lock();
+  // If inheritance not handled properly, this unlock might not match the lock
+  // above because technically they act on different memory regions:
+  // __mutex_base and mutex.
+  m.unlock();
+  sleep(10); // no-warning
+}

